helix-core/src/syntax.rs

```diff
@@ -107,8 +110,13 @@ pub struct LanguageConfiguration {
     #[serde(skip)]
     pub(crate) highlight_config: OnceCell<Option<Arc<HighlightConfiguration>>>,
     // tags_config OnceCell<> https://github.com/tree-sitter/tree-sitter/pull/583
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub language_server: Option<LanguageServerConfiguration>,
+    #[serde(
+        default,
+        skip_serializing_if = "HashMap::is_empty",
+        serialize_with = "serialize_lang_features",
+        deserialize_with = "deserialize_lang_features"
+    )]
+    pub language_servers: HashMap<String, LanguageServerFeatures>,
```

Comment:
While, refactoring `Editor::launch_language_servers` (as noted by @pascalkuthe here: https://github.com/helix-editor/helix/pull/2507#discussion_r1141223543) I noticed, that now with the new `HashMap` instead of a `Vec`, the order of the language servers in a language-configuration is not guaranteed anymore, which is necessary for a correct fallback to the next specified language-server.
I think we need something like a `Vec<(String, LanguageServerFeatures)>` here instead, to be able to keep the order, that the user specified in the configuration.

A `Vec` also enables something like this (not that it makes sense in most practical contexts I guess, but this should be able to be handled correctly IMHO with this syntax):

The following would only enable formatting for "lsp1" use everything else from "lsp2" and if "lsp2" doesn't support a feature, fallback to "lsp1" if possible:

```toml
[[language]]
language-servers = [
  { name = "lsp1", only-features = "format" },
  "lsp2",
  "lsp1"
]
```

So either we completely rethink how this is handled (especially with the context of merging project specific configurations as noted here: https://github.com/helix-editor/helix/pull/2507#discussion_r1137416399), or we think about how this syntax could be integrated meaningfully with merging lsp-configurations (either global injection as below, or project-wide configuration overrides)
An interesting aspect is also global LSP injection as noted by @pickfire here: https://github.com/helix-editor/helix/pull/2507#discussion_r1139570723

Which would also make things a little bit more complicated.

Does anyone have suggestions here, so that in the future we don't have to completely change the configuration structure/logic (better now, before it's "stable")?

Reply:
I think the config structures as we have now is fine it's just a matter of deserializing correctly. In this case the right data structure would be an [`indexmap`](https://docs.rs/indexmap/latest/indexmap/). You can handroll this using a `hashbrown::raw` but in this case that's probably not a good API to expose. I think just pulling in `indexmap` is fine here since we already depend on it trough `gitoxide`

Reply:
Ah I see your point about the fallback behavior. altough your old implementation didn't support that either. The problem is that for most features (like diagnostics) we only care if a certain feature is enabled for a server.  I looked into  the fallback behavior again and I think the `language-servers` array would be overwritten instead of merged anyway (we stop merging at depth 3) so that's ok.


It's probably not that big of a deal to do a linear search on a `Vec` wherever you do a `HashMap` lookup right now. However to make this eplicit fallback order thing actually work you need to turn the `langauge_servers` field into a
`HashMap<LanguageServerName, Arc<Client>>` and then iterate the configured language_servers (filtered by whether they support the requested capability) and then use the name from the configuration to obtain the actual client.

This would actually play well with the `launch_server` problem too because you can then simply compute `HashMap` diffs there (`helix_lsp::Registry::get` should also return a `HashMap` then).

Reply:
> altough your old implementation didn't support that either

It did, if I'm not mistaken? They were launched in the order of the language-specific language-server config, so they retained the order. Then they were just filtered by features and following that by LSP capabilities for the requested LSP request.

It could be that there was buggy/non-desired behavior non-the-less (like duplicated completions because the language server may be iterated twice, not the case with the previous example, but probably for a slightly different example). I guess I have to think a few cases through more carefully, maybe write some tests for different behaviors, if possible/feasible.

Anyway, sounds like a good idea, then I keep the current behavior, and we're targeting complete lsp-configuration overrides for project-specific configuration, similar as the current user-default-configuration-overrides merge behavior.

I guess global LSP injections also only really make sense for features, that can be merged (completion/diagnostics code-actions etc.) and not for "single-request" (go-to-reference etc.). Maybe for this use-case just give them the lowest priority (i.e. equivalent to last `Vec` entry in the language-specific language-server configuration)

Reply:
Just one more thing: Please still keep the custom deserializer I suggested so that we don't expose the `enum` to the rest of the editor. You can just convert the current `HashMap<String, LanguageServerFeatures>` to `Vec<LangaugeServerFeatures>` and add a `name` field.

> It could be that there was buggy/non-desired behavior non-the-less (like duplicated completions because the language server may be iterated twice, not the case with the previous example, but probably for a slightly different example). I guess I have to think a few cases through more carefully, maybe write some tests for different behaviors, if possible/feasible.


Thee easiest thing to do would be to use a `HashSet` for de-duplication in `Document::language_servers_with_features` (just `.filter(move |ls| seen_ls.insert(ls))`)

Reply:
Thanks for the tip with the `HashSet`, I did fix now all the things here AFAIK.

Reply:
Yeah that's better 👍